go/vt/vtgate/engine/batch_route.go

```diff
@@ -0,0 +1,126 @@
+	panic("call not expected")
+}
+
+// ExecuteBatch performs a non-streaming exec of list of queries.
+func (batchRoute *BatchRoute) ExecuteBatch(vcursor VCursor, queryBatchConstruct *queryinfo.QueryBatchConstruct, joinvars map[string]interface{}, wantfields bool) ([]sqltypes.QueryResponse, error) {
+	switch batchRoute.Opcode {
+	case ExecuteOrdered:
+		return batchRoute.executeOrdered(vcursor, queryBatchConstruct, joinvars, wantfields)
+	case ExecuteUnOrdered:
+		return batchRoute.executeUnordered(vcursor, queryBatchConstruct, joinvars, wantfields)
+	default:
+		return nil, fmt.Errorf("ExecuteBatch:unsupported query route: %v", batchRoute)
+	}
+}
+
+func (batchRoute *BatchRoute) executeOrdered(vcursor VCursor, queryBatchConstruct *queryinfo.QueryBatchConstruct, joinvars map[string]interface{}, wantfields bool) ([]sqltypes.QueryResponse, error) {
+	// AsTransaction flag is discarded.
+	queryResponses := make([]sqltypes.QueryResponse, len(queryBatchConstruct.BoundQueryList))
+	for i, plan := range batchRoute.PlanList {
+		if plan == nil {
+			queryResponses[i] = sqltypes.QueryResponse{
+				QueryResult: nil,
+				QueryError:  errors.New("no plan generated for this query"),
+			}
+			continue
+		}
+		queryConstruct := queryinfo.NewQueryConstruct(queryBatchConstruct.BoundQueryList[i].SQL, queryBatchConstruct.Keyspace, queryBatchConstruct.BoundQueryList[i].BindVars, false)
+		result, err := plan.Instructions.Execute(vcursor, queryConstruct, joinvars, wantfields)
+		queryResponses[i] = sqltypes.QueryResponse{
+			QueryResult: result,
+			QueryError:  err,
+		}
+	}
+	return queryResponses, nil
+}
+
+type response struct {
+	queryResponse sqltypes.QueryResponse
+	queryIndex    int
+}
+
+func (batchRoute *BatchRoute) executeUnordered(vcursor VCursor, queryBatchConstruct *queryinfo.QueryBatchConstruct, joinvars map[string]interface{}, wantfields bool) ([]sqltypes.QueryResponse, error) {
+	// AsTransaction flag is discarded.
+	var wg sync.WaitGroup
+
+	queryResponses := make([]sqltypes.QueryResponse, len(queryBatchConstruct.BoundQueryList))
+	ch := make(chan response)
+
+	for index, plan := range batchRoute.PlanList {
+		wg.Add(1)
+		go func(index int, plan *Plan) {
+			defer wg.Done()
+			if plan == nil {
+				ch <- response{
+					queryIndex: index,
+					queryResponse: sqltypes.QueryResponse{
+						QueryResult: nil,
+						QueryError:  errors.New("no plan generated for this query"),
+					},
+				}
+				return
+			}
+			queryConstruct := queryinfo.NewQueryConstruct(queryBatchConstruct.BoundQueryList[index].SQL, queryBatchConstruct.Keyspace, queryBatchConstruct.BoundQueryList[index].BindVars, false)
+			result, err := plan.Instructions.Execute(vcursor, queryConstruct, joinvars, wantfields)
+			ch <- response{
+				queryIndex:    index,
+				queryResponse: sqltypes.QueryResponse{QueryResult: result, QueryError: err},
+			}
+		}(index, plan)
+
+	}
+	go func() {
```

Comment:
Found the bug: you need to wait for this function to also complete before returning.
Also, this function will currently never return because you're not closing ch anywhere.